README.md
LocationPicker
==============

LocationPicker is a Google Maps widget to help with entering geographic data
into web forms. It provides an address search field to place a marker on the
map; the marker can then be dragged to the exact location required, firing
events that update form fields with the marker's coordinates.


Usage
-----

LocationPicker uses version 2 of the Google Maps API.

### Loading the Google Maps library

If loading through Helium, the Google API key must be set:

    Helium.GOOGLE_API_KEY = '{your API key}';

Your [Helium custom loader file][helium] will need to include some code to load
the Google Maps library:

    /**
     * Loads the `google.load` function, required to load other
     * parts of the Google API. Requires `Helium.GOOGLE_API_KEY`
     * to be set beforehand.
     **/
    loader(function(cb) {
        var url = 'http://www.google.com/jsapi?key=' + Helium.GOOGLE_API_KEY;
        load(url, cb);
    })
    .provides('google.load');
    
    /**
     * Loads the Google Maps API. Requires `Helium.GOOGLE_API_KEY`
     * to be set beforehand.
     **/
    loader(function(cb) { google.load('maps', '2.x', {callback: cb}) })
        .provides('GMap2',  'GClientGeocoder',
                  'GEvent', 'GLatLng', 'GMarker')
        .requires('google.load');

Otherwise, load the Google Maps library as specified in the
[Google Maps API V2 documentation][gmap2docs].

### Using `LocationPicker`

To use the `LocationPicker` class you'll need some HTML elements to plug it
into.

    <form id="dataForm">
        <p>
            <label for="lat">Latitude</label>
            <input id="lat" name="lat" type="text" value="">
        </p>
        
        <p>
            <label for="lng">Latitude</label>
            <input id="lng" name="lng" type="text" value="">
        </p>
    </form>

Instantiate the widget by specifying the dimensions of the map and references to the
form fields that should be updated when the marker is moved:

    new LocationPicker('after', '#dataForm', {
        width:      600,
        height:     400,
        latField:   '#lat',
        lngField:   '#lng'
    });

The first two arguments specify where to insert the widget in the document;
here we insert it `after` the element selected by `#dataForm`. The other valid
position arguments are `before`, `top` and `bottom`.

  [helium]:    http://github.com/othermedia/helium
  [gmap2docs]: http://code.google.com/apis/maps/documentation/javascript/v2/index.html


API methods
-----------

Instances of the `LocationPicker` class have the following methods available.

### `LocationPicker#getHTML()`

Returns the HTML representation of the widget.

### `LocationPicker#pullLocation()`

Extracts location data from the form fields and updates the marker location.

### `LocationPicker#pushLocation()`

Updates the form fields with the current marker position.

### `LocationPicker#getMarker()`

Returns the Google Maps marker used by the widget.

### `LocationPicker#positionMarker()`

Places the marker at the correct location and updates the form fields.

### `LocationPicker#search([address])`

Runs a geocoding search using the current address value and moves the marker to
the resulting location. If the `address` parameter is provided, the address
field will be set to that value prior to performing the search.

### `LocationPicker#getAddress()`

Returns the value of the address field.

### `LocationPicker#setAddress(address)`

Sets the value attribute of the address field to the `address` argument.


CHANGELOG
Version 1.0.2, 2010.08.31

* Fix test case form label.

Version 1.0.1, 2010.08.31

* Fix erroneous object reference in getAddress method.

Version 1.0.0, 2010.08.31

* LocationPicker, a Google Maps widget to help with entering geographic data
  into web forms


test/packages.js
JS.Packages(function() { with(this) {

    file('./lib/location-picker-min.js')
        .provides("LocationPicker")
        .requires("JS.MethodChain", "JS.Class", "GMap2", "GClientGeocoder", "GEvent", "GLatLng", "GMarker")
        .uses("Ojay", "Ojay.HTML");
}});


source/location-picker.js
// IMPORTANT - DO NOT REMOVE
if (window.GUnload) Ojay(window).on('unload', GUnload);
if (window.GMap2) JS.MethodChain.addMethods(GMap2);

/**
 * == libs ==
 **/

/** section: libs
 * class LocationPicker
 * 
 * The `LocationPicker` class provides a UI for entering geographic data
 * (lat/lng co-ordinates) into web forms. The user can search for an address
 * to position a map marker, then drag the marker to the precise location
 * they want. Interaction fires events that can export the current marker
 * position to form fields.
 **/
LocationPicker = new JS.Class({
    extend: {
        // Default dimensions
        WIDTH:              600,
        HEIGHT:             400,
        
        // HTML classes
        CONTAINER_CLASS:    'location-picker',
        
        // Default location: tOm office
        DEFAULT_LATITUDE:   51.498945,
        DEFAULT_LONGITUDE:  -0.080874,
        
        // Default zoom level and co-ordinate precision
        // 6 decimal places in lat/lng values = 11cm on-the-ground precision
        ZOOM_LEVEL:         15,
        PRECISION:          6,
        
        /**
         * LocationPicker.round(number) -> Number
         * - number (Number)
         * Returns the given number rounded to the current precision level.
         **/
        round: function(number) {
            var scale = Math.pow(10, this.PRECISION);
            return Math.round(number * scale) / scale;
        },
        
        /**
         * LocationPicker.getGeocoder() -> GClientGeocoder
         * Returns a geocoder object from Google Maps.
         **/
        getGeocoder: function() {
            return this._geoc = this._geoc || new GClientGeocoder();
        }
    },
    
    /**
     * new LocationPicker(position, element, options)
     * - position (String)
     * - element (HTMLElement)
     * - options (Object)
     * 
     * `LocationPicker` is initialized using a `position` (either `'before'` or
     * `'after'`), and an element after/before which to insert to map UI. The `options`
     * object should include:
     * 
     * * `width` -- width of the widget in pixels
     * * `height` -- height of the widget in pixels
     * * `latField` -- reference to the form field to populuate with latitude
     * * `lngField` -- reference to the form field to populuate with longitude
     **/
    initialize: function(position, element, options) {
        this._container = Ojay( Ojay.HTML.div({className: this.klass.CONTAINER_CLASS}) );
        this._options = options || {};
        this._elements = {};
        this._latField = Ojay(this._options.latField).node;
        this._lngField = Ojay(this._options.lngField).node;
        
        Ojay(element).insert(this.getHTML(), position);
        this._map = new GMap2(this._elements._map.node);
        this._map.enableScrollWheelZoom();
        this._map.enableContinuousZoom();
        this._map.disableDoubleClickZoom();
        this._map.addControl(new GSmallMapControl());
        
        GEvent.addListener(this._map, 'dblclick', function(overlay, latlng) {
            this._loc = latlng;
            this.positionMarker();
        }.bind(this));
        
        this.pullLocation();
    },
    
    /**
     * LocationPicker#getHTML() -> Ojay.DomCollection
     * Returns the HTML that makes up the widget.
     **/
    getHTML: function() {
        var elements = this._elements;
        if (elements._container) return elements._container;
        
        elements._container = Ojay( Ojay.HTML.div({className: this.klass.CONTAINER_CLASS}, function(HTML) {
            elements._form = Ojay( HTML.form(function(HTML) {
                elements._address = HTML.input({type: 'text'});
                elements._button = Ojay( HTML.button('Search') );
            }) ); 
            elements._map = Ojay( HTML.div() );
        }) );
        
        elements._map.setStyle({
            width:    (this._options.width || this.klass.WIDTH) + 'px',
            height:   (this._options.height || this.klass.HEIGHT) + 'px'
        });
        
        elements._form.on('submit', Ojay.stopDefault)._(this).search();
        
        return elements._container;
    },
    
    /**
     * LocationPicker#pullLocation() -> undefined
     * Extracts location data from the form fields and updates the marker location.
     **/
    pullLocation: function() {
        var lat = (this._latField||{}).value, lng = (this._lngField||{}).value;
        if (!lat) lat = this.klass.DEFAULT_LATITUDE;
        if (!lng) lng = this.klass.DEFAULT_LONGITUDE;
        
        this._loc = new GLatLng(Number(lat), Number(lng));
        this._map.setCenter(this._loc, this.klass.ZOOM_LEVEL);
        this.positionMarker();
    },
    
    /**
     * LocationPicker#pushLocation() -> undefined
     * Updates the form fields with the current marker position.
     **/
    pushLocation: function() {
        var latlng = this._marker.getLatLng();
        (this._latField||{}).value = this.klass.round(latlng.lat());
        (this._lngField||{}).value = this.klass.round(latlng.lng());
    },
    
    /**
     * LocationPicker#getMarker() -> GMarker
     * Returns the Google Maps marker used by the widget.
     **/
    getMarker: function() {
        if (this._marker) return this._marker;
        this._marker = new GMarker(this._loc, {draggable: true});
        this._marker.enableDragging();
        GEvent.addListener(this._marker, 'dragend', this.method('pushLocation'));
        this._map.addOverlay(this._marker);
        return this._marker;
    },
    
    /**
     * LocationPicker#positionMarker() -> undefined
     * Places the marker at the correct location and updates the form fields.
     **/
    positionMarker: function() {
        this.getMarker().setLatLng(this._loc);
        this.pushLocation();
    },
    
    /**
     * LocationPicker#search() -> undefined
     * - address (String)
     * Runs a geocoding search using the current address value and moves the
     * marker to the resulting location. If the address parameter is provided,
     * the address field will be set to that value prior to performing the
     * search.
     **/
    search: function(address) {
        if (address) this.setAddress(address);
        
        this.klass.getGeocoder().getLocations(this.getAddress(), function(response) {
            var place = response.Placemark[0];
            if (!place) return;
            
            var coords = place.Point.coordinates,
                box    = place.ExtendedData.LatLonBox,
                bounds = new GLatLngBounds(new GLatLng(box.south, box.west),
                                           new GLatLng(box.north, box.east)),
                zoom = this._map.getBoundsZoomLevel(bounds);
            
            this._loc = new GLatLng(coords[1], coords[0]);
            this._map.setCenter(this._loc, zoom);
            this.positionMarker();
        }.bind(this));
    },
    
    /**
     * LocationPicker#getAddress() -> String
     * Returns the value of the address field.
     */
    getAddress: function() {
        return this._elements._address.value;
    },
    
    /**
     * LocationPicker#setAddress(value) -> undefined
     * - value (String)
     * Sets the value of the address field.
     **/
    setAddress: function(value) {
        this._elements._address.value = value;
    }
});
